src/systems/logical_audio_sensor_plugin/LogicalAudioSensorPlugin.cc

```diff
@@ -0,0 +1,499 @@
+      this->dataPtr->playSource = false;
+    }
+    play_lock.unlock();
+
+    // stop playing a source if the stop source service was called
+    std::unique_lock<std::mutex> stop_lock(this->dataPtr->stopSourceMutex);
+    if (this->dataPtr->stopSource)
+    {
+      playInfo.playing = false;
+      this->dataPtr->stopSource = false;
+    }
+    stop_lock.unlock();
+
+    // stop playing a source if the play duration has been exceeded
+    if (this->dataPtr->DurationExceeded(_info, playInfo))
+      playInfo.playing = false;
+  }
+}
+
+//////////////////////////////////////////////////
+void LogicalAudioSensorPlugin::PostUpdate(const UpdateInfo &_info,
+                const EntityComponentManager &_ecm)
+{
+  // check to see which sources a microphone can hear
+  if (this->dataPtr->micEntity != kNullEntity)
+  {
+    const auto micPose = worldPose(this->dataPtr->micEntity, _ecm);
+    const auto micInfo = _ecm.Component<components::LogicalMicrophone>(
+        this->dataPtr->micEntity)->Data();
+
+    _ecm.Each<components::LogicalAudioSource,
+              components::LogicalAudioSourcePlayInfo>(
+      [&](const Entity &_entity,
+          const components::LogicalAudioSource *_source,
+          const components::LogicalAudioSourcePlayInfo *_playInfo)
+      {
+        // skip this source if the playing duration has been exceeded
+        // (this source will be stopped in the following PreUpdate call)
+        if (this->dataPtr->DurationExceeded(_info, _playInfo->Data()))
+          return true;
+
+        const auto sourcePose = worldPose(_entity, _ecm);
+        const auto vol = logical_audio::ComputeVolume(
+            _playInfo->Data().playing,
+            _source->Data().attFunc,
+            _source->Data().attShape,
+            _source->Data().emissionVolume,
+            _source->Data().innerRadius,
+            _source->Data().falloffDistance,
+            sourcePose,
+            micPose);
+
+        if (logical_audio::Detect(vol, micInfo.volumeDetectionThreshold))
+        {
+          // TODO(adlarkin) let others know that a detection occurred,
+          // perhaps by publishing a message to a topic?
+
+          // using a static mutex temporarily so that it's easy to
+          // read console output (in case multiple microphones can detect
+          // a source in the same time step)
+          static std::mutex write_mutex;
+          std::lock_guard<std::mutex> lock(write_mutex);
+          ignmsg << "microphone " << micInfo.id << " detected source "
+            << _source->Data().id << " at sim time " << _info.simTime.count()
+            << "\n";
+        }
+
+        return true;
+      });
+  }
+}
+
+//////////////////////////////////////////////////
+void LogicalAudioSensorPluginPrivate::CreateAudioSource(
+    const sdf::ElementPtr &_elem,
+    const Entity &_parent,
+    EntityComponentManager &_ecm,
+    SdfEntityCreator &_sdfEntityCreator)
+{
+  static std::unordered_set<unsigned int> sourceIDs;
```

Comment:
Yes, that's a good idea - thanks for pointing that out (I used a `static set` originally because I forgot that I could use the ECM instead). I went ahead and incorporated this in https://github.com/ignitionrobotics/ign-gazebo/commit/f35c133ef7e22c432ee2db786f74c9a814a03bab. 

I've got a follow up question about this: since I am checking for duplicate IDs in `System::Configure` (that's where source and microphone creation occurs), is there ever a chance that multiple instances of the same plugin can have their `System::Configure` call be run in parallel? If so, I should add in a mutex for the ECM so that one thread isn't trying to check for duplicate microphone/source entities while another thread is trying to add a new microphone/source entity.

Reply:
As a follow up, it looks like the `System::Configure` calls are NOT ran in parallel (https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/src/SimulationRunner.cc#L760), so I think this can be marked as resolved.